pkg/volume/azure_dd/attacher.go

```diff
@@ -0,0 +1,289 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure_dd
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/arm/compute"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/keymutex"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type azureDiskAttacher struct {
+	host          volume.VolumeHost
+	azureProvider azureCloudProvider
+}
+
+var _ volume.Attacher = &azureDiskAttacher{}
+
+var _ volume.AttachableVolumePlugin = &azureDataDiskPlugin{}
+
+const (
+	checkSleepDuration = time.Second
+)
+
+// Azure doesn't allow simultanious disk attach or detach on the same instance
+// serialize concurrent attach/detach per instance
+var attachDetachMutex = keymutex.NewKeyMutex()
+
+func (plugin *azureDataDiskPlugin) NewAttacher() (volume.Attacher, error) {
+	azure, err := getAzureCloudProvider(plugin.host.GetCloudProvider())
+	if err != nil {
+		glog.V(4).Infof("failed to get azure provider")
+		return nil, err
+	}
+
+	return &azureDiskAttacher{
+		host:          plugin.host,
+		azureProvider: azure,
+	}, nil
+}
+
+func (attacher *azureDiskAttacher) Attach(spec *volume.Spec, hostName string) (string, error) {
+	volumeSource, err := getVolumeSource(spec)
+	if err != nil {
+		glog.V(2).Infof("failed to get azure disk spec")
+		return "", err
+	}
+	instanceid, err := attacher.azureProvider.InstanceID(hostName)
+	if err != nil {
+		glog.V(2).Infof("failed to get azure instance id")
+		return "", fmt.Errorf("failed to get azure instance id for host %q", hostName)
+	}
+	if ind := strings.LastIndex(instanceid, "/"); ind >= 0 {
+		instanceid = instanceid[(ind + 1):]
+	}
+	attachDetachMutex.LockKey(instanceid)
+	defer attachDetachMutex.UnlockKey(instanceid)
+
+	lun, err := attacher.azureProvider.GetDiskLun(volumeSource.DiskName, volumeSource.DataDiskURI, instanceid)
+	if err == cloudprovider.InstanceNotFound {
```

Comment:
Sure. But my point is if `err != nil && err != cloudprovider.InstanceNotFound` then what happens? It's not logged?